### PR TITLE
Allow using C++23

### PR DESCRIPTION
**Pull request by masterleinad**
This still assumes that C++14 is allowed.

### CODE REVIEW

**Review by crtrott — APPROVED** [review]
Looks good to me

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp [hunk 1]
```diff
@@ -458,12 +458,7 @@ class ParallelReduce<FunctorType, Kokkos::MDRangePolicy<Traits...>, ReducerType,
   using pointer_type   = typename Analysis::pointer_type;
   using reference_type = typename Analysis::reference_type;
 
-  enum {
-    HasJoin =
-        Impl::FunctorAnalysis<Impl::FunctorPatternInterface::REDUCE, Policy,
-                              FunctorType>::has_join_member_function
-  };
-  enum { UseReducer = is_reducer<ReducerType>::value };
+  static constexpr bool UseReducer = is_reducer<ReducerType>::value;
```
Comment (nliber) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp:461]: This change doesn't look to be part of "Allow using C++23".

**Review by nliber — CHANGES_REQUESTED** [review]

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp:461]: It's for avoiding warnings building with C++20. This pull request also adds better support for C++20 and thus also tests it.
Once, this pull request actually passes CI, I am happy to discuss splitting it further if that's what people want.

**Comment by masterleinad** [comment]
This is good enough for me now. To have at least one C++20 and one C++23 it seemed easiest to pick CI builds that require very recent compilers anyway. I first played with `OpenMPTarget` and `SYCL` and ran into toolchain issues with the GCC headers. Bumping the HIP builds, though, seemed to work fine.
I still kept the `OpenMPTarget` changes since we will be running into those warnings anyway.

**Comment by masterleinad** [comment]
The `OPENACC-NVHPC-CUDA-11.6` build should be fixed by #5288.

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp:461]: Please let me know if you insist on dropping these changes.

**Review by dalg24 — CHANGES_REQUESTED** [review]
I object to the CI changes as they are now
Bumping 14 -> 17 and 17 -> 20, sure.  But not C++23

File: cmake/kokkos_pick_cxx_std.cmake [hunk 2]
```diff
@@ -20,6 +21,12 @@ IF (KOKKOS_CXX_STANDARD)
   ELSEIF(${KOKKOS_CXX_STANDARD} STREQUAL "c++17")
     MESSAGE(WARNING "Deprecated Kokkos C++ standard set as 'c++17'. Use '17' instead.")
     SET(KOKKOS_CXX_STANDARD "17")
+  ELSEIF(${KOKKOS_CXX_STANDARD} STREQUAL "c++20")
+    MESSAGE(WARNING "Deprecated Kokkos C++ standard set as 'c++20'. Use '20' instead.")
+    SET(KOKKOS_CXX_STANDARD "20")
+  ELSEIF(${KOKKOS_CXX_STANDARD} STREQUAL "c++23")
+    MESSAGE(WARNING "Deprecated Kokkos C++ standard set as 'c++23'. Use '23' instead.")
+    SET(KOKKOS_CXX_STANDARD "23")
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply cmake/kokkos_pick_cxx_std.cmake:29]: Let's discuss that in another pull request.

[the remark this answers is not in the document]
Comment (crtrott) [reply cmake/kokkos_pick_cxx_std.cmake:29]: yeah I do want one C++23 build, but its fine with me if that is a GCC build on github. using GCC 12 preferrably.

[the remark this answers is not in the document]
Comment (crtrott) [reply cmake/kokkos_pick_cxx_std.cmake:29]: This can happenlater btw.

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp (hunk 1 above)
Comment (nliber) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp:461]: I'm fine with it.

**Review by nliber — APPROVED** [review]

File: core/src/Kokkos_Macros.hpp [hunk 3]
```diff
@@ -124,7 +124,8 @@
 #define KOKKOS_LAMBDA [=]
 #endif
 
-#if (defined(KOKKOS_ENABLE_CXX17) || defined(KOKKOS_ENABLE_CXX20)) && \
+#if (defined(KOKKOS_ENABLE_CXX17) || defined(KOKKOS_ENABLE_CXX20) || \
+     defined(KOKKOS_ENABLE_CXX23)) &&                                \
```
Comment (dalg24) [inline core/src/Kokkos_Macros.hpp:128]: Mark it `FIXME_CXX17`

File: core/src/Kokkos_Macros.hpp [hunk 4]
```diff
@@ -644,7 +645,8 @@ static constexpr bool kokkos_omp_on_host() { return false; }
 #define KOKKOS_ENABLE_CUDA_LDG_INTRINSIC
 #endif
 
-#if defined(KOKKOS_ENABLE_CXX17) || defined(KOKKOS_ENABLE_CXX20)
+#if defined(KOKKOS_ENABLE_CXX17) || defined(KOKKOS_ENABLE_CXX20) || \
+    defined(KOKKOS_ENABLE_CXX23)
```
Comment (dalg24) [inline core/src/Kokkos_Macros.hpp:649]: Same comment

File: core/src/setup/Kokkos_Setup_Cuda.hpp [hunk 5]
```diff
@@ -84,7 +84,8 @@
 #ifdef KOKKOS_ENABLE_CUDA_LAMBDA
 #define KOKKOS_LAMBDA [=] __host__ __device__
 
-#if defined(KOKKOS_ENABLE_CXX17) || defined(KOKKOS_ENABLE_CXX20)
+#if defined(KOKKOS_ENABLE_CXX17) || defined(KOKKOS_ENABLE_CXX20) || \
+    defined(KOKKOS_ENABLE_CXX23)
```
Comment (dalg24) [inline core/src/setup/Kokkos_Setup_Cuda.hpp:88]: FIXME_CXX17

File: core/src/Kokkos_Macros.hpp (hunk 3 above)
Comment (masterleinad) [reply core/src/Kokkos_Macros.hpp:128]: All these places are already (supposed to be part of https://github.com/kokkos/kokkos/issues/5295) and I would assume any cleanup after requiring C++17 to search for `KOKKOS_ENABLE_CXX17` anyway.

**Pull request merged**